src/os/bluestore/Writer.cc

```diff
@@ -0,0 +1,1384 @@
+    if (left_b != emap.end()) {
+      uint32_t in_blob_offset = logical_offset - left_b->blob_start();
+      uint32_t in_blob_end = in_blob_offset + bd_it->disk_data.length();
+      uint32_t data_end_offset = logical_offset + bd_it->disk_data.length();
+      _maybe_expand_blob(left_b->blob.get(), p2roundup(in_blob_end, au_size));
+      _blob_put_data_subau_allocate(
+        left_b->blob.get(), in_blob_offset, bd_it->disk_data);
+      uint32_t ref_end = std::min(ref_end_offset, data_end_offset);
+      //fixme/improve - need something without stupid extras - that is without coll
+      left_b->blob->get_ref(coll, in_blob_offset, ref_end - logical_offset);
+      Extent *le = new Extent(
+        logical_offset, in_blob_offset, ref_end - logical_offset, left_b->blob);
+      dout(20) << __func__ << " new extent " << le->print(pp_mode) << dendl;
+      emap.insert(*le);
+      logical_offset = ref_end;
+      ++bd_it;
+    } else {
+      // it is still possible to use first bd and put it into
+      // blob after punch_hole
+      // can blob before punch_hole be different then blob after punch_hole ?
+    }
+  }
+  if (bd_it != bd.end()) {
+    // still something to process
+    auto back_it = bd.end() - 1;
+    if (!back_it->is_compressed()) {
+      // it is thinkable to put the data to some after
+      uint32_t left_bound = p2align(data_end_offset, blob_size);
+      uint32_t right_bound = p2roundup(data_end_offset, blob_size);
+      exmp_it right_b = _find_mutable_blob_right(
+          after_punch_it, left_bound, right_bound,
+          data_end_offset - back_it->disk_data.length(), data_end_offset);
+      if (right_b != emap.end()) {
+        // before putting last blob, put previous
+        // it is nicer to have AUs in order.
+        _do_put_new_blobs(logical_offset, ref_end_offset, bd_it, back_it);
+        uint32_t data_begin_offset = data_end_offset - back_it->disk_data.length();
+        uint32_t in_blob_offset = data_begin_offset - right_b->blob_start();
+        _maybe_expand_blob(right_b->blob.get(), in_blob_offset + bd_it->disk_data.length());
+        _blob_put_data_subau_allocate(
+          right_b->blob.get(), in_blob_offset, back_it->disk_data);
+        uint32_t ref_end = std::min(ref_end_offset, data_begin_offset + back_it->disk_data.length());
+        //fixme - need something without stupid extras
+        right_b->blob->get_ref(coll, in_blob_offset, ref_end - data_begin_offset);
+        Extent *le = new Extent(
+          data_begin_offset, in_blob_offset, ref_end - data_begin_offset, right_b->blob);
+        dout(20) << __func__ << " new extent " << le->print(pp_mode) << dendl;
+        emap.insert(*le);
+        bd.erase(back_it); //TODO - or other way of limiting end
+      }
+    }
+  }
+
+  // that's it about blob reuse, now is the time to full blobs
+  if (bd_it != bd.end()) {
+    _do_put_new_blobs(logical_offset, ref_end_offset, bd_it, bd.end());
```

Comment:
I find it hard to understand this part. I imagine it's something like:

1. if we can reuse first blob, do it
2. if we c an reuse last blob, do it
3. process all other blobs

If that's correct why is the range of blobs in 3): bd_it up to bd.end() instead of, bd_it up to bd.end() - 1?

Reply:
I would prefer having conditions like 
```c++
if (first_blob_exists) {
    try_to_reuse_it...
} 
if (last_blob_exits) {
    try_to_reuse_it..
}

if (blobs_left_to_process > 0) {
    do_put_new_blobs
}
```

Reply:
`_blob_put_data_subau_allocate` and `_do_put_new_blobs` are consuming allocations prepared by `_defer_or_allocate`.
So, if we have a continuous 10 AUs, and invoke:
- try_to_reuse (first)
- try_to_reuse (last)
- put_new_blobs
We will get order of AUs in object: 0 2 3 4 5 6 7 8 9 1.
I didn't like it.